implement checking for pattern match exhaustiveness and for unreachable cases

zygoloid opened the pull request:
When checking for control flow falling off a function after a `match`, determine whether it's possible for no case to have matched. Using the same implementation, also detect whether `case`s in a `match` are unreachable.

For now, the implementation never considers a match against specific values for any type other than tuples, alternatives, and `bool` to be exhaustive. In particular, matching against the sole value `{}` of type `{}` is not considered exhaustive. This is probably best left until explorer supports matching on struct and maybe class types more generally.

This implementation closely follows the algorithm described in the paper [Warnings for pattern matching](http://moscova.inria.fr/~maranget/papers/warn/warn.pdf) by Luc Maranget. Various optimizations are possible, such as reducing the amount of copying done, but for the purposes of explorer, comprehensibility is being favored over efficiency.

The problem is, perhaps surprisingly, co-NP-hard (by reduction to the tautology problem for disjunctive normal form, which is in turn dual to the satisfaction problem for conjunctive normal form, which is well-known to be NP-hard). The algorithm is therefore exponential-time in the worst case, but seems to be well-studied and performs well enough on non-pathological examples.

zygoloid committed:
First pass at proper exhaustive match checking.

zygoloid committed:
Decompose compound values, clean up, add tests.

zygoloid committed:
Rename SimplePattern -> AbstractPattern to be a bit clearer about *how*
it's been simplified from an AST pattern.

chandlerc commented:
> The algorithm is therefore exponential-time in the worst case, but seems to be well-studied and performs well enough on non-pathological examples.

Just a quick side-question -- is there some limiting in place to avoid this becoming a place that fuzzers explore and get stuck "discovering" pathological cases? If not, is that feasible?

Mostly want to preserve our ability to fuzz teh explorer more generally w/o getting lost cleverly discovering timeouts here.

jonmeow commented:
> > The algorithm is therefore exponential-time in the worst case, but seems to be well-studied and performs well enough on non-pathological examples.
> 
> Just a quick side-question -- is there some limiting in place to avoid this becoming a place that fuzzers explore and get stuck "discovering" pathological cases? If not, is that feasible?
> 
> Mostly want to preserve our ability to fuzz teh explorer more generally w/o getting lost cleverly discovering timeouts here.

The answer may be to limit how much complexity the structured fuzzing can output for match cases. But it may also be fine to wait and see what issues arise for timeouts.

jonmeow commented on explorer/interpreter/type_checker.cpp:
```diff
@@ -2920,6 +2920,354 @@ auto TypeChecker::TypeCheckPattern(
+    Primitive
+  };
+
+  AbstractPattern(Nonnull<const Pattern*> pattern) { Set(pattern); }
```
```suggestion
  explicit AbstractPattern(Nonnull<const Pattern*> pattern) { Set(pattern); }
```

zygoloid replied:
It's important for the ergonomics elsewhere that this conversion is implicit, eg filling a vector of `AbstractPattern`s from a range of `const Pattern*`s.

jonmeow replied:
Can you then add `// NOLINTNEXTLINE(google-explicit-constructor)` and an explanation?

zygoloid replied:
Done.

jonmeow commented on explorer/interpreter/type_checker.cpp:
```diff
@@ -2920,6 +2920,354 @@ auto TypeChecker::TypeCheckPattern(
+
+  // Determine whether the given pattern vector is useful: that is, whether
+  // adding it to the matrix would allow any more values to be matched.
+  auto IsUseful(llvm::ArrayRef<AbstractPattern> pattern) const -> bool {
```
It's hard for me to understand what's "useful". Maybe `IsRedundant` returning the opposite would be easier to understand? `CanHaveMatches`?

zygoloid replied:
This name comes from the paper. Happy to change if you would still prefer a different name (and perhaps the opposite sense) given that fact.

jonmeow replied:
I think my preference would be a more understandable name because most people won't read the paper. :)

zygoloid replied:
I've made the public API be an `IsRedundant` function, and moved `IsUseful` alongside the other private member functions that are named after concepts from the paper.

jonmeow commented on explorer/interpreter/type_checker.cpp:
```diff
@@ -2920,6 +2920,354 @@ auto TypeChecker::TypeCheckPattern(
+      }
+      return Primitive;
+    }
+    assert(value_.is<const WildcardTag*>());
```
Should your `assert` uses generally be `CARBON_CHECK`?

zygoloid replied:
Done.

jonmeow commented on explorer/interpreter/type_checker.cpp:
```diff
@@ -2920,6 +2920,354 @@ auto TypeChecker::TypeCheckPattern(
+    }
+  }
+
+ private:
```
Multiple private sections?

zygoloid replied:
Yeah, this is an ingrained habit from a non-Google coding style to have different `private` sections for member functions versus data. I'll try to stop :)

```suggestion
```

jonmeow commented on explorer/interpreter/type_checker.cpp:
```diff
@@ -2920,6 +2920,354 @@ auto TypeChecker::TypeCheckPattern(
+  }
+
+ private:
+  struct alignas(8) WildcardTag {};
```
Why does this need `alignas`? (comment?)

zygoloid replied:
```suggestion
  // This is aligned so that we can use it in the `PointerUnion` below.
  struct alignas(8) WildcardTag {};
```

jonmeow commented on explorer/interpreter/type_checker.cpp:
```diff
@@ -2920,6 +2920,354 @@ auto TypeChecker::TypeCheckPattern(
+          break;
+        }
+        case AbstractPattern::Primitive: {
+          // TODO: We assume that primitive value matches are always incomplete
```
Can you add a simple test with `case 1`, `case 2`, etc just to make sure it's working as intended?

Maybe do something with `case 1` repeated?

zygoloid replied:
Done.

jonmeow commented on explorer/interpreter/type_checker.cpp:
```diff
@@ -2920,6 +2920,354 @@ auto TypeChecker::TypeCheckPattern(
+    PatternMatrix default_matrix;
+    for (auto& row : matrix_) {
+      assert(!row.empty());
+      switch (row[0].kind()) {
```
Why the `switch` here versus `if`? Catch what happens if more are added?

zygoloid replied:
This was mostly to make this function parallel `Specialize` / `SpecializeRow` as much as possible since it's doing something quite similar. I'm not worried about warnings for new cases given that we really only care about wildcard cases here. Happy to change to an `if` if you'd prefer?

jonmeow commented on explorer/interpreter/type_checker.cpp:
```diff
@@ -2920,6 +2920,354 @@ auto TypeChecker::TypeCheckPattern(
+    return default_matrix;
+  }
+
+ private:
```
Repeating private again? Might be easier to read this with a forward declaration.

zygoloid replied:
Will fix repeated `private`.

I didn't follow what you would like to be forward-declared here, can you elaborate?

jonmeow replied:
I mean the higher level "separate class member declarations and definitions" so that it'd be easier to see the repeat `private` on a single screen. :)

zygoloid replied:
I'll do that when I move this to a separate file.

jonmeow replied:
👍

zygoloid replied:
Done.

jonmeow reviewed (commented):
Is this something that it might make sense to switch out to a separate file? type_checker.cpp is already pretty long, and this sort of looks like a clean unit of code that could be in its own file.

Note, I'm not looking closely at the paper to see how it compares to the implementation.

zygoloid committed:
Add limit to prevent exponential behavior.

Limit is approximately tuned so that Series T from the reference paper
takes <0.1s in a development build on my workstation, but this will vary
depending on the nature of the pathological testcase presented.

This could be made much faster (and the limit correspondingly increased)
with a little work; the implementation is currently *very* naive.

jonmeow commented on explorer/interpreter/type_checker.cpp:
```diff
@@ -2920,6 +2920,372 @@ auto TypeChecker::TypeCheckPattern(
   }
 }
 
+// A abstracted view of a pattern or constant value (which we view as a
```
```suggestion
// An abstracted view of a pattern or constant value (which we view as a
```

zygoloid commented:
> Is this something that it might make sense to switch out to a separate file? type_checker.cpp is already pretty long, and this sort of looks like a clean unit of code that could be in its own file.

Will do once other review comments are resolved a little, I think this would mess up github's comment threads.

> > The algorithm is therefore exponential-time in the worst case, but seems to be well-studied and performs well enough on non-pathological examples.
> 
> Just a quick side-question -- is there some limiting in place to avoid this becoming a place that fuzzers explore and get stuck "discovering" pathological cases? If not, is that feasible?

I've added an explicit depth limit to the exponential step for now; I guess we'll see if people hit that in practice, but you'd need quite a complex pattern to get there. I don't think it's all that plausible that the current fuzzer will be able to create complex cases here, due to the need to simultaneously increase the complexity of both the scrutinee and the pattern and have them result in the same type in order to produce a new testcase that would even reach this code. But maybe it's smarter than I give it credit for, and maybe we'll have a more semantics-aware fuzzer in the future.

jonmeow commented on explorer/interpreter/type_checker.cpp:
```diff
@@ -2920,6 +2920,372 @@ auto TypeChecker::TypeCheckPattern(
+  // recursively expanding the pattern. Allowing this to happen an arbitrary
+  // number of times leads to exponential growth in the runtime of the
+  // algorithm.
+  static constexpr int MaxExponentialDepth = 8;
```
Note, this should be at the top of the class declaration order because it's a static constant.

zygoloid replied:
Done.

zygoloid committed:
Apply suggestions from code review

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

zygoloid committed:
Code review comments.

zygoloid committed:
Fix unit test failure cause by incorrect detection of 'default' case.

zygoloid committed:
Rename IsUseful to IsRedundant (and reverse sense).

zygoloid committed:
Add test for unreachable `default`.

zygoloid committed:
Factor pattern analysis code out into its own header and source file.

zygoloid committed:
Support exhaustiveness checks for pattern matches on `bool`.

jonmeow commented on explorer/interpreter/type_checker.cpp:
```diff
@@ -3075,17 +3075,38 @@ class AbstractPattern {
+  }
+
+  // Is the given pattern vector redundant if it appears after the patterns in
+  // this matrix? That is, will it never match following the other patterns?
```
```suggestion
  // Returns true if the given pattern vector is redundant when it appears after the patterns in
  // this matrix. That is, will it never match following the other patterns?
```

Comments should clearly describe behavior to the reader; only asking questions may leave readers wondering who's supposed to answer.
https://google.github.io/styleguide/cppguide.html#Function_Comments

jonmeow commented on explorer/testdata/match/fail_exhaustive_int.carbon:
```diff
@@ -0,0 +1,23 @@
+    case 1 => { return 1; }
+    case 2 => { return 2; }
+  }
+// CHECK: COMPILATION ERROR: {{.*}}/explorer/testdata/match/fail_exhaustive_int.carbon:[[@LINE+1]]: non-exhaustive match may allow control-flow to reach the end of a function that provides a `->` return type
```
I admit, this makes me want a couple tests that should pass:

```
fn A(n: i32) -> i32 {
  match (n) {
    case 1 => { return 1; }
    case 2 => { return 2; }
  }
  return 3;
}
```

```
fn A(n: i32) {
  match (n) {
    case 1 => { return; }
    case 2 => { return; }
  }
}
```

zygoloid replied:
Sure, sounds reasonable. I think these are tests for the "control reaches end of function" diagnostic, not tests for the pattern match exhaustiveness checking code, so I've added them to testdata/function/.

jonmeow commented on explorer/interpreter/pattern_analysis.cpp:
```diff
@@ -0,0 +1,307 @@
+  return {};
+}
+
+auto AbstractPattern::NumElements() const -> int {
```
FWIW, I don't understand why `kind()` or `discriminator()` accessor naming, but then `NumElements` here (they seem to have similar characteristics). I don't have a strong opinion one way or the other, but it may make sense to change one or the other.

zygoloid replied:
Yeah, I find this awkward too. `kind` and `discriminator` are trying to follow the "use lowercase for property names" rule. It's not clear to me whether `NumElements` should follow that rule too or not, but it doesn't feel especially property-like to me, so I used the "normal" CamelCase convention.

zygoloid replied:
Renamed to `elements_size` per offline discussion.

jonmeow reviewed (approved):
Basically LGTM, just a few last comments (please address them as you like, I don't feel strongly about the resolutions -- though as I write that, please do rephrase the one comment-asking-the-reader).

zygoloid committed:
Add some additional requested tests for the "control reaches end of
function" diagnostic.

zygoloid committed:
Replace `not` with `%{not}` in tests for portability.

zygoloid committed:
Replace question in function description with statement.

zygoloid committed:
Rename NumElements to element_size.
